File: OST Tracker/ViewControllers/OSTUtilitiesViewController.m
//
//  OSTUtilitiesViewController.m
//  OST Remote
//
//  Created by Guillermo Apoj on 3/19/20.
//  Copyright © 2020 OST. All rights reserved.
//

#import "OSTUtilitiesViewController.h"
#import "OSTEventSelectionViewController.h"
#import "AppDelegate.h"
#import "UIView+Additions.h"
#import "CurrentCourse.h"
#import "EffortModel.h"

@interface OSTUtilitiesViewController ()
@property (weak, nonatomic) IBOutlet UILabel *lblTitle;
@property (strong, nonatomic) IBOutlet UIView *loadingView;
@property (weak, nonatomic) IBOutlet UILabel *lblYourDataIsSynced;
@property (weak, nonatomic) IBOutlet UIImageView *imgCheckMark;
@property (weak, nonatomic) IBOutlet UILabel *lblSuccess;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (weak, nonatomic) IBOutlet UIProgressView *progressBar;
@property (weak, nonatomic) IBOutlet UIButton *btnReturnToLiveEntry;
@property (weak, nonatomic) IBOutlet UILabel *lblSyncing;
@property (weak, nonatomic) IBOutlet UIImageView *logoImage;
@property (weak, nonatomic) IBOutlet UILabel *remoteLbl;
@property (weak, nonatomic) IBOutlet UIButton *btnRetry;

@end

@implementation OSTUtilitiesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (IS_IPHONE_X || IS_IPHONE_XR)
    {
        self.lblTitle.numberOfLines = 1;
        self.lblTitle.bottom = self.lblTitle.bottom + 7;
        self.menuButton.bottom = self.menuButton.bottom + 7;
        
    }
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    self.loadingView.size = self.view.size;
}

- (void) showLoadingScreen
{
    [self showLoadingValues];
    self.loadingView.size = self.view.size;
    [self.view addSubview:self.loadingView];
    [self.view bringSubviewToFront:self.loadingView];
    self.loadingView.alpha = 0;
    
    __weak OSTUtilitiesViewController * weakSelf = self;
    [UIView animateWithDuration:0.5 animations:^{
        weakSelf.loadingView.alpha = 1;
    }];
}

- (void) showLoadingValues
{
    self.btnRetry.hidden = YES;
    self.imgCheckMark.hidden = YES;
    self.lblSuccess.hidden = YES;
    self.lblYourDataIsSynced.hidden = YES;
    self.btnReturnToLiveEntry.hidden = YES;
    
    [self.activityIndicator startAnimating];
    
    self.progressBar.hidden = NO;
    self.logoImage.hidden = NO;
    self.remoteLbl.hidden = NO;
    
}

- (void) showFinishLoadingValues
{
    self.imgCheckMark.image = [UIImage imageNamed:@"CheckMark"];
    self.imgCheckMark.hidden = NO;
    self.lblSuccess.text = @"Success!";
    self.lblSuccess.hidden = NO;
    self.lblYourDataIsSynced.text = @"The entrants data has been updated";
    self.lblYourDataIsSynced.hidden = NO;
    self.btnReturnToLiveEntry.hidden = NO;
   
    
    [self.activityIndicator stopAnimating];
    self.lblSyncing.hidden = YES;
    
    self.progressBar.hidden = YES;
    self.logoImage.hidden = YES;
    self.remoteLbl.hidden = YES;
}

- (void) showFinishLoadingErrorValues:(NSString*) error
{
    self.imgCheckMark.image = [UIImage imageNamed:@"Error-icon"];
    self.imgCheckMark.hidden = NO;
    self.lblSuccess.text = @"Failure!";
    self.lblSuccess.hidden = NO;
    self.lblYourDataIsSynced.text = error;
    self.lblYourDataIsSynced.hidden = NO;
    self.btnReturnToLiveEntry.hidden = NO;
    self.btnRetry.hidden = NO;
    
    [self.activityIndicator stopAnimating];
    self.lblSyncing.hidden = YES;
    
    self.progressBar.hidden = YES;
    self.logoImage.hidden = YES;
    self.remoteLbl.hidden = YES;
}
- (IBAction)onRefreshData:(id)sender {
    [self showLoadingScreen];
    self.progressBar.progress = 0.5;
    CurrentCourse * currentCourse = [CurrentCourse getCurrentCourse];
    
    __weak OSTUtilitiesViewController * weakSelf = self;
    [[AppDelegate getInstance].getNetworkManager getEventsDetails:currentCourse.eventId completionBlock:^(id object)
     {
        weakSelf.progressBar.progress = 1;
        [weakSelf.activityIndicator stopAnimating];
        
        
        currentCourse.dataEntryGroups = object[@"data"][@"attributes"][@"dataEntryGroups"];
        for (id dataObject in object[@"included"])
        {
            if ([dataObject[@"type"] isEqualToString:@"efforts"])
            {
                [EffortModel MR_importFromObject:dataObject];
            }
        }
        //        Remove from local storage entrants that were deleted from the event 
        NSArray *idList = [object[@"included"] valueForKey:@"id"];
        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"NOT(effortId IN %@)", idList];
        [EffortModel MR_deleteAllMatchingPredicate:predicate];
        
        
        currentCourse.monitorPacers = object[@"data"][@"attributes"][@"monitorPacers"];
        
        NSMutableDictionary * eventIdsAndSplits = [NSMutableDictionary new];
        NSMutableDictionary * eventShortNames = [NSMutableDictionary new];
        for (NSDictionary * dict in object[@"included"])
        {
            if ([dict[@"type"] isEqualToString:@"events"])
            {
                NSString *shortName = dict[@"attributes"][@"shortName"];
                if (shortName != nil && [shortName isKindOfClass:[NSString class]])
                {
                    eventShortNames[dict[@"id"]] = shortName;
                }
                NSMutableArray * arr = eventIdsAndSplits[dict[@"id"]];
                if (arr == nil)
                {
                    arr = [NSMutableArray new];
                }
                [arr addObject: dict[@"attributes"][@"parameterizedSplitNames"]];
                eventIdsAndSplits[dict[@"id"]] = arr;
            }
        }
        currentCourse.eventIdsAndSplits = eventIdsAndSplits;
        currentCourse.eventShortNames = eventShortNames;
         [UIView animateWithDuration:0.5 animations:^{
           [weakSelf showFinishLoadingValues];
         }];
        [[NSManagedObjectContext MR_defaultContext] processPendingChanges];
        [[NSManagedObjectContext MR_defaultContext] MR_saveOnlySelfAndWait];
        
        
    } errorBlock:^(NSError *error) {
       
       [UIView animateWithDuration:0.5 animations:^{
           [weakSelf showFinishLoadingErrorValues:error.localizedDescription];
        }];
    }];
}


- (IBAction)onReturnToLiveEntry:(id)sender
{
    [self.activityIndicator stopAnimating];
    self.loadingView.hidden = YES;
    [self.loadingView removeFromSuperview];
    [[AppDelegate getInstance] showTracker];
    
}

- (IBAction)onAbout:(id)sender {
    [[AppDelegate getInstance] showAbout];
}

- (IBAction)onMenu:(id)sender
{
    [[AppDelegate getInstance].rightMenuVC toggleRightSideMenuCompletion:nil];
}

- (IBAction)onChangeStation:(id)sender
{
    OSTEventSelectionViewController * event = [[OSTEventSelectionViewController alloc] initWithNibName:nil bundle:nil];
    event.changeStation = YES;
    [self presentViewController:event animated:YES completion:nil];
}

- (IBAction)onLogout:(id)sender
{
    if (![AppDelegate getInstance].getNetworkManager.reachabilityManager.reachable)
    {
        [OHAlertView showAlertWithTitle:@"Logout is disabled" message:@"Please try again when you have an Internet connection" cancelButton:@"Ok" otherButtons:nil buttonHandler:nil];
        return;
    }
    
    [OHAlertView showAlertWithTitle:@"Are you sure you would like to log out?" message:@"You will not be able to log back in or add entries until you have a data connection again." cancelButton:@"Cancel" otherButtons:@[@"Logout"] buttonHandler:^(OHAlertView *alert, NSInteger buttonIndex) {
        if (buttonIndex == 1)
        {
            [[AppDelegate getInstance].rightMenuVC toggleRightSideMenuCompletion:nil];
            [[AppDelegate getInstance] logout];
        }
    }];
}

@end

File: OST Tracker/ViewControllers/OSTUtilitiesViewController.swift
//
//  OSTUtilitiesViewController.swift
//  OST Tracker
//
//  Migrated from Objective-C (Phase 1). Keeps the existing XIB; OHAlertView
//  replaced with native UIAlertController. Still uses the Obj-C network manager
//  and MagicalRecord via bridging (migrated in later phases).
//

import UIKit
import CoreData

@objc(OSTUtilitiesViewController)
class OSTUtilitiesViewController: OSTBaseViewController {

    @IBOutlet weak var lblTitle: UILabel!
    @IBOutlet var loadingView: UIView!
    @IBOutlet weak var lblYourDataIsSynced: UILabel!
    @IBOutlet weak var imgCheckMark: UIImageView!
    @IBOutlet weak var lblSuccess: UILabel!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    @IBOutlet weak var progressBar: UIProgressView!
    @IBOutlet weak var btnReturnToLiveEntry: UIButton!
    @IBOutlet weak var lblSyncing: UILabel!
    @IBOutlet weak var logoImage: UIImageView!
    @IBOutlet weak var remoteLbl: UILabel!
    @IBOutlet weak var btnRetry: UIButton!

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        ostApplySafeAreaFix()
        ostPositionBadgeAtMenu()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        loadingView.frame.size = view.frame.size
    }

    // MARK: - Loading states

    private func showLoadingScreen() {
        showLoadingValues()
        loadingView.frame.size = view.frame.size
        view.addSubview(loadingView)
        view.bringSubviewToFront(loadingView)
        loadingView.alpha = 0
        UIView.animate(withDuration: 0.5) { self.loadingView.alpha = 1 }
    }

    private func showLoadingValues() {
        btnRetry.isHidden = true
        imgCheckMark.isHidden = true
        lblSuccess.isHidden = true
        lblYourDataIsSynced.isHidden = true
        btnReturnToLiveEntry.isHidden = true
        activityIndicator.startAnimating()
        progressBar.isHidden = false
        logoImage.isHidden = false
        remoteLbl.isHidden = false
    }

    private func showFinishLoadingValues() {
        imgCheckMark.image = UIImage(named: "CheckMark")
        imgCheckMark.isHidden = false
        lblSuccess.text = "Success!"
        lblSuccess.isHidden = false
        lblYourDataIsSynced.text = "The entrants data has been updated"
        lblYourDataIsSynced.isHidden = false
        btnReturnToLiveEntry.isHidden = false
        activityIndicator.stopAnimating()
        lblSyncing.isHidden = true
        progressBar.isHidden = true
        logoImage.isHidden = true
        remoteLbl.isHidden = true
    }

    private func showFinishLoadingErrorValues(_ error: String) {
        imgCheckMark.image = UIImage(named: "Error-icon")
        imgCheckMark.isHidden = false
        lblSuccess.text = "Failure!"
        lblSuccess.isHidden = false
        lblYourDataIsSynced.text = error
        lblYourDataIsSynced.isHidden = false
        btnReturnToLiveEntry.isHidden = false
        btnRetry.isHidden = false
        activityIndicator.stopAnimating()
        lblSyncing.isHidden = true
        progressBar.isHidden = true
        logoImage.isHidden = true
        remoteLbl.isHidden = true
    }

    // MARK: - Actions

    @IBAction func onRefreshData(_ sender: Any) {
        showLoadingScreen()
        progressBar.progress = 0.5
        guard let currentCourse = CurrentCourse.getCurrentCourse() else { return }

        OSTBackend.shared.getEventsDetails(currentCourse.eventId ?? "") { [weak self] object, error in
            guard let self = self else { return }
            if let error = error {
                UIView.animate(withDuration: 0.5) { self.showFinishLoadingErrorValues(error.localizedDescription) }
                return
            }
            self.progressBar.progress = 1
            self.activityIndicator.stopAnimating()

            let root = object as? [String: Any]
            let attributes = (root?["data"] as? [String: Any])?["attributes"] as? [String: Any]
            currentCourse.dataEntryGroups = attributes?["dataEntryGroups"]

            let included = root?["included"] as? [[String: Any]] ?? []
            EffortModel.mr_reconcile(fromIncluded: included, ofType: "efforts")
            currentCourse.monitorPacers = attributes?["monitorPacers"] as? NSNumber

            var eventIdsAndSplits = [String: [Any]]()
            var eventShortNames = [String: String]()
            for dict in included where (dict["type"] as? String) == "events" {
                guard let eventId = dict["id"] as? String else { continue }
                let attrs = dict["attributes"] as? [String: Any]
                if let shortName = attrs?["shortName"] as? String {
                    eventShortNames[eventId] = shortName
                }
                var arr = eventIdsAndSplits[eventId] ?? []
                if let psn = attrs?["parameterizedSplitNames"] { arr.append(psn) }
                eventIdsAndSplits[eventId] = arr
            }
            currentCourse.eventIdsAndSplits = eventIdsAndSplits
            currentCourse.eventShortNames = eventShortNames

            UIView.animate(withDuration: 0.5) { self.showFinishLoadingValues() }
            NSManagedObjectContext.mr_default().processPendingChanges()
            NSManagedObjectContext.mr_default().mr_saveOnlySelfAndWait()
        }
    }

    @IBAction func onReturnToLiveEntry(_ sender: Any) {
        activityIndicator.stopAnimating()
        loadingView.isHidden = true
        loadingView.removeFromSuperview()
        AppDelegate.getInstance()?.showTracker()
    }

    @IBAction func onAbout(_ sender: Any) {
        AppDelegate.getInstance()?.showAbout()
    }

    @IBAction func onMenu(_ sender: Any) {
        AppDelegate.getInstance()?.rightMenuVC.toggleRightSideMenuCompletion(nil)
    }

    @IBAction func onChangeStation(_ sender: Any) {
        let event = OSTEventSelectionViewController(nibName: nil, bundle: nil)
        event.changeStation = true
        present(event, animated: true)
    }

    @IBAction func onLogout(_ sender: Any) {
        let checking = UIAlertController(title: "Checking connection…",
                                         message: "\n\n",
                                         preferredStyle: .alert)
        let spinner = UIActivityIndicatorView(style: .gray)
        spinner.translatesAutoresizingMaskIntoConstraints = false
        spinner.startAnimating()
        checking.view.addSubview(spinner)
        NSLayoutConstraint.activate([
            spinner.centerXAnchor.constraint(equalTo: checking.view.centerXAnchor),
            spinner.bottomAnchor.constraint(equalTo: checking.view.bottomAnchor, constant: -16)
        ])
        present(checking, animated: true)

        OSTBackend.shared.verifyConnection { [weak self] error in
            guard let self = self else { return }
            checking.dismiss(animated: true) {
                if error == nil {
                    self.presentLogoutConfirmation()
                } else {
                    self.presentLogoutOverride()
                }
            }
        }
    }

    /// Online: confirm, then log out.
    private func presentLogoutConfirmation() {
        let alert = UIAlertController(title: "Are you sure you would like to log out?",
                                      message: "You can log back in using your current connection, but you won't be able to add entries or log back in if you lose it.",
                                      preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        alert.addAction(UIAlertAction(title: "Logout", style: .destructive) { [weak self] _ in
            self?.performLogout()
        })
        present(alert, animated: true)
    }

    /// Check failed: block, but allow an immediate override.
    private func presentLogoutOverride() {
        let alert = UIAlertController(title: "Can't reach OpenSplitTime",
                                      message: "You will not be able to log back in or add entries until you have a data connection again.",
                                      preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        alert.addAction(UIAlertAction(title: "Log Out Anyway", style: .destructive) { [weak self] _ in
            self?.performLogout()
        })
        present(alert, animated: true)
    }

    private func performLogout() {
        let app = AppDelegate.getInstance()
        app?.rightMenuVC.toggleRightSideMenuCompletion(nil)
        app?.logout()
    }
}
